BUG: Fix slowness in stats.mode

gclenaghan opened the pull request:
Currently, the main thread of stats.mode is O(size of array * number of distinct elements), which is often O(N^2). Confusingly, if `nan_policy='omit'` and there is at least one `nan`, then it uses a completely different implementation which is fast. I mostly followed that implementation with one twist to allow any type.

I also spotted another bug: the omit implementation returned the first element and not the smallest when there were no repeated items. I added a test for this.

I also noticed that `nan_policy='propagate'` behaves exactly the same as `nan_policy='omit'`. There was actually a test for this (test_stats.py line 1398) so I left that alone for now but I'm pretty sure something is wrong here. I believe after `_contains_nan` it should just return early. This is an easy add but I wanted to ask before messing around with old tests as a new contributor. Most importantly, it probably breaks a lot of people's code.

Fixes #1432

tylerjereddy commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1402,6 +1402,21 @@ def test_mode_nan(self):
         assert_raises(ValueError, stats.mode, data1, nan_policy='raise')
         assert_raises(ValueError, stats.mode, data1, nan_policy='foobar')
 
+    def test_smallest_equal(self):
```
Some concision might be gained by writing a [parametrized test](https://docs.pytest.org/en/latest/parametrize.html) here, with i.e., a "data" parameter which corresponds to each of the data lists, and then a single result & assert statement. One possible drawback is that you'd also likely need another parameter for `nan_policy` specification, and could end up running a few more test permutations by default, but that might be ok.

gclenaghan commented:
Thanks for the feedback @tylerjereddy. I've made the changes you suggest. I don't need to parametrize `nan_policy` as with it set to omit it chooses the code path based on whether it contains a `nan` or not.

The benchmark takes 40ms in the released scipy, 1.2ms after these changes.

One thing I think I need feedback on is the behavior in Python 2: right now there is a test that fails in 2 and was already marked as xfail in Python 3. [The test is here](https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_stats.py#L1342-L1352).

Basically, the issue is that the new behavior relies on `np.unique` to count instances, in this case of an array of objects of various types, including `nan`. `np.unique` already fails for mixed types in Python 3 thus the xfail. In Python 2, it doesn't throw any exceptions, but silently returns bad results since it does not sort `nan` correctly. In the implementation I'm replacing, it worked since it did not rely on `np.unique` to actually return a unique list of items. Either way it fails (even before these changes) if you pass `nan_policy='omit'` since that code path uses `scipy.stats.find_repeats` which fails on on non-numeric arrays.

I don't think there is an elegant solution to handling this edge case, and I'm not sure it's worth adding an inelegant solution given that it's already accepted to fail in Python 3. I lean towards removing or changing the test but I'd like to hear your thoughts first.

Thanks!

tylerjereddy commented on benchmarks/benchmarks/stats.py:
```diff
@@ -92,5 +92,15 @@ def time_distribution(self, distribution, properties):
+
+
+class DescriptiveStats(Benchmark):
+    def setup(self):
```
Probably clearer to [parametrize](http://asv.readthedocs.io/en/latest/writing_benchmarks.html#parameterized-benchmarks) here -- you can make setup accept an argument for the size of the data set to be used and then the tests will be categorized accordingly with their performance. i.e., setup(self, n_points).

tylerjereddy commented on benchmarks/benchmarks/stats.py:
```diff
@@ -92,5 +92,15 @@ def time_distribution(self, distribution, properties):
+class DescriptiveStats(Benchmark):
+    def setup(self):
+        np.random.seed(12345678)
+        self.many_levels = np.random.randint(1000, size=(1000, 10))
```
Pretty sure the random seed will be consumed on the first random call -- instead of adding a second fixed seed, probably easier to just parametrize the data sizes and keep a single fixed seed and a single i.e., self.levels type call instead of two separate ones.

tylerjereddy commented:
One of the `scipy.stats` gurus should also review & may have some deeper thoughts on handling dtype=object scenarios and the np.unique issue. We generally have a pretty strong obligation to backward compatibility so I certainly can't recommend breaking that without more input from others.

gclenaghan commented:
Hi @tylerjereddy 

I forgot about this for a while, but I think I've resolved the remaining issues. I parametrized the benchmark as you suggested. I also added a fallback logic: if the dtype is object and the array contains a NaN, it uses the previous (slow) implementation, with the only change being that it uses `set` instead of `np.unique` as the latter does not handle NaNs in this case. It also falls back to this if the version of Numpy is older than 1.9, which does not have `return_counts` in `np.unique`. The change to `set` also now makes this work in Python 3, so I removed the xfail annotation on the test.

I also rebased it on the latest master branch.

Let me know what you think.

gclenaghan commented:
Thanks. Is there someone that would be good to reach out to? Not sure about the process for this.

rgommers commented:
LGTM too, merged. Thanks @gclenaghan and @tylerjereddy 

> Not sure about the process for this.

Asking on the issue is usually the best. We're somewhat under-resourced so we don't always manage to review within a week.
